### PR TITLE
WIP/RFC: REPL: Fuzzy Fish-like tab completion

### PR DESCRIPTION
**Pull request by ghost**
Firstly, let me just state that this is an outright filthy implementation. But I think an implementation is necessary to experience what is proposed to better form an opinion.

A little less than a year ago I took the [fish shell](https://fishshell.com) for a spin to experience a “modern” shell. I liked what it brought and soon I left Bash behind after more than a decade. One feature that I really came to love is the [fuzzy matching it incorporates into its tab completion](https://github.com/fish-shell/fish-shell/issues/568#issuecomment-18455181). While this initially may not seem like that much of a deal – at least it did not to me – it very elegantly allows tab completion to work both as one would expect with a more “naive” shell or REPL and also as a very quick way to dig into and search a big namespace. Thus when I got back into hacking Julia more seriously about a month ago I could not help feeling a bit frustrated to have to resort to either the browser or the Levenshtein-based search in the REPL help mode.

I “borrowed” what I think is the current algorithm in fish (GPL… unable to read code…) and made a very dirty quick hack that only works with the module namespace (it most likely breaks for a number of cases given how tired I was when writing this). The algorithm uses the following priority, returning once there is one or more matches for a given step:

1. Exact match
2. Prefix match
3. Case insensitive exact match
4. Case insensitive prefix match
5. Substring match
6. Subsequence match

Despite seeming somewhat simple, this strategy strikes some sort of balance between expected and “advanced” behaviour. Below are some examples from a few test cases, but I really encourage you to take this for a spin as it really took me by surprise initially how useful it was at fish and scoffed at it when reading about it.

```julia
const STRINGS = ["foobar", "foo", "bar", "baz", "qux", "quux", "quuz"]

@test fuzzyfilter("foo", STRINGS) == ["foo"]
@test fuzzyfilter("fo", STRINGS) == sort(["foobar", "foo"])
@test fuzzyfilter("BAR", STRINGS) == ["bar"]
@test fuzzyfilter("BA", STRINGS) == sort(["bar", "baz"])
@test fuzzyfilter("ux", STRINGS) == sort(["qux", "quux"])
@test fuzzyfilter("qx", STRINGS) == sort(["qux", "quux"])
@test fuzzyfilter("foobaz", STRINGS) == []
```

If this is ever to be integrated, it will need a much cleaner implementation. But given the complexity of the REPL code and how much of an effort it would be, I thought it wise to float the idea first for comments. To the best of my knowledge and looking through previous PRs and issues, fuzzy completions have not been proposed in the past.

Lastly, please ignore the NixOS-specific commit in the PR. It is just there to make development less of a pain given my OS, it would be removed prior to any merge and should have no effect on your end.

**Edit:** Grammar

### CODE REVIEW

**Comment by ninjin** [comment]
Already found a few bugs playing around (*sigh*), but one has to admit that this is pretty elegant:

```julia
julia> ception
CapturedException      ErrorException          InterruptException      MissingException        TaskFailedException
CompositeException     Exception               InvalidStateException   ProcessFailedException
```

**Comment by aviatesk** [comment]
fwiw, there's already a package to provide fuzzy completioes: https://github.com/JunoLab/FuzzyCompletions.jl

I _think_ fuzzy completions don't necessarily live in Julia stdlib since the functionality can be implemented as a third party package, and making REPL completions fuzzy by default is opinionated.
What do you think on this ?

**Comment by KristofferC** [comment]
Just as a reference, fuzzy finding in REPL history  (not tab completion) is available with https://github.com/Moelf/JLFzf.jl or OhMyREPL.jl.

**Comment by ninjin** [comment]
> fwiw, there's already a package to provide fuzzy completioes: https://github.com/JunoLab/FuzzyCompletions.jl

Wow! Thanks a ton @aviatesk! Few things make as happy as seeing an implementation of something that I want already out there.

> I _think_ fuzzy completions don't necessarily live in Julia stdlib since the functionality can be implemented as a third party package, and making REPL completions fuzzy by default is opinionated.
> What do you think on this ?

I agree that it does not necessarily have to be included in base (I hesitate to call it stdlib as it is not as if the user really calls the REPL as a library). However, a common complaint about Julia is in discoverability, as we naturally have a much larger namespace compared to languages such as Python and also find it more difficult to fall back on types as many static languages do. From my perspective, this causes undue friction for beginners and immigrants alike, but my hope is that one could alleviate this by technical means rather than despair as multiple dispatch is a wonderful thing indeed.

Allow me to elaborate, in a dynamic language most experimental coding will take place in the REPL. What I mean by this, is that you are likely to dive into the REPL to build a small prototype or just confirm your understanding of a specific piece of logic. Tab completion in its vanilla form really accomplishes two things, it saves you having to type out the entirety of a long name that can be uniquely identified by the first initial characters and allows for a “cheap” kind of search for names that you can make a qualified guess for (think `is` -> `isa isabspath isabstracttype…`). It is this latter part I think where fuzzy matching really can shine.

However, if the defining characteristics of whatever you are looking for resides towards the end of a name – which it usually does – tab completion in its vanilla form is of little help to you. Similarly, it makes it fairly cumbersome to “guess” a part of what you are looking for as you would need to work from left to right and it is insufficient to just get a noun (or a portion of it) right towards the end. Not to mention that you must get casing right and avoid misspellings.

For the above, fuzzy matching – I believe – can be extremely useful. Imagine for example using it to navigate our [massive space of Unicode symbols](https://docs.julialang.org/en/v1/manual/unicode-input) (I know I have to go to that page in the manual and do a few text searches whenever I can not remember a symbol) or just as I demonstrated above use it to fish out all exceptions in the current namespace. Most of the time, this can save you a trip to your search engine of choice. At least if you use the fish-derived algorithm, you do retain the excepted vanilla behaviour while (maybe) surprisingly never really getting surprised by what you get back after pressing tab. The only things I would add is that it would be helpful to highlight the matched portions (we already do this for search) and possibly incorporate a completion selection algorithm akin to fish and zsh if you push tab a second time.

So, in conclusion, I would do not think it is overly opinionated and that a patch based on FuzzyCompletions.jl could be a wonderful thing to empower our users to find the beauty in large namespaces. One should probably take it for a lengthy trial run first though, but from what little I have toyed around with my dirty patch so far, it does work pretty much exactly as one would want it to.

> Just as a reference, fuzzy finding in REPL history (not tab completion) is available with https://github.com/Moelf/JLFzf.jl or OhMyREPL.jl.

Thanks @KristofferC, I did spot that one! Not sure exactly how I would stand on fzf-style history in base – although I love it in my shell… I suppose it would require an fzf implementation in Julia though.

**Comment by ninjin** [comment]
All right, patch reworked and should work just fine for module, property, and field completions now. An implementation for backslash, shell, and path completions should not be too hard, but I think to take this for a spin the three we currently have implemented are enough. Anyone willing to take it for a spin? How is the user experience in your book?

@aviatesk: One thing I realised while writing this, I think `FuzzyCompletions.jl` is very different in that it just falls back on fuzzy as opposed to strict prefix matching. What we have here is:

1. Exact match
2. Prefix match
3. Case insensitive exact match
4. Case insensitive prefix match
5. Substring match
6. Case insensitive substring match
7. Subsequence match

Which gives you a very different user experience.

@pfitzseb: If you do not mind me asking, what was the story behind [your original fuzzy matching patch](https://github.com/pfitzseb/julia/commit/740dd16843c16cb0b87264911f43abf8485652fe)?

**Comment by pfitzseb** [comment]
> @pfitzseb: If you do not mind me asking, what was the story behind your original fuzzy matching patch?

From what I can remember I didn't like the user experience, mostly because fuzzy matching absolutely needs to allow the user to select the match from the displayed list -- narrowing it down by typing more is very unergonomic or even downright impossible. And I couldn't really be bothered writing that UI at the time ;) Should maybe take this PR as a reason to re-think that decision.

Regarding the algorithm: https://github.com/jeancroy/fuzz-aldrin-plus works really well imho, in case you want to try comparing your implementation with something MIT licensed (and the readme is worth a read regardless).

**Comment by ninjin** [comment]
On Wed 12 May 2021, Sebastian Pfitzner wrote:
> 
> > @pfitzseb: If you do not mind me asking, what was the story behind your original fuzzy matching patch?
> 
> From what I can remember I didn't like the user experience, mostly because fuzzy matching absolutely needs to allow the user to select the match from the displayed list -- narrowing it down by typing more is very unergonomic or even downright impossible. And I couldn't really be bothered writing that UI at the time ;) Should maybe take this PR as a reason to re-think that decision.

I think I agree with your conclusion here, a selection mechanism is more or less necessary to comfortably complete non-prefix matches. If you (or anyone else) have the cycles and want to take a stab at it I would love to have the code. Otherwise I will probably try to take a stab at it at some point, although it could take another week given my other FLOSS/work commitments.

On that note, if someone wants to take ownership off me at any point and push for this feature you are more than welcome. Ultimately what I want to accomplish with this is not to prove my coding skills or have more patches in `Base` and I would gladly hand it off and take more of a backseat role.

> Regarding the algorithm: https://github.com/jeancroy/fuzz-aldrin-plus works really well imho, in case you want to try comparing your implementation with something MIT licensed (and the readme is worth a read regardless).

No kidding, thank you, that was a good read! I think I will stay the course with the fish-based algorithm for now (one advantage is that it is O(n) with a hefty constant rather than O(n²)). Most of the issues they raise I do think are addressed by the steps of the current algorithm, but I should revisit it later on as I do like the prospect of a single simple algorithm rather than a chain of simple ones. On that matter, any more reading on this is more than welcome if you guys have pointers.

Lastly, if there is any immediate opposition to this kind of feature I would really love to hear it before pouring more hours into this. This is not to say that I do not enjoy de-rusting my Julia skills, but there are only so many hours in the day and maybe I am better off supporting the goals of our community elsewhere? So far I have mostly seen encouragement and I would love to have a patch in before the 1 June feature freeze for v1.7.x so that I can brag about our REPL being advanced when I give my classes in Julia next Autumn.

**Comment by ninjin** [comment]
Closing in favour of https://github.com/JuliaLang/julia/pull/43838 which is closer to completion and with substantially cleaner code.
